### PR TITLE
[ts2pant-du-tagged-encoding] Patch 1: Structural discriminated-union detection + tagged-encoding synth (dormant)

### PR DESCRIPTION
**Pull request by subsetpark**
## Patch 1: Structural discriminated-union detection + tagged-encoding synth (dormant)

- Implement detectDiscriminatedUnion structurally, reading each member's discriminant literal value; bail (null) on any non-discriminated shape. When multiple fields qualify as discriminants, pick the first by sorted field-name order (deterministic, total).
- Add the DiscriminatedUnionSynth family (type, empty, SynthCell field, newSynthCell init, cellRegisterDiscriminatedUnion, emitDiscriminatedUnionSynthDecls) mirroring the Map/Record synth; the discriminant rule is unguarded, variant-field rules carry the discriminant-equality guard as a PantRule.guard OpaqueExpr.
- Drain the new synth in cellEmitSynth.
- Add detection + emission unit tests and the fixtures (positive + negative); register the positives in the allowlist; add skipped end-to-end stubs.
- Confirm no emitted-document change (the new path is unconsumed by mapTsType/field access).

## Changes
- Implement detectDiscriminatedUnion structurally, reading each member's discriminant literal value; bail (null) on any non-discriminated shape. When multiple fields qualify as discriminants, pick the first by sorted field-name order (deterministic, total).
- Add the DiscriminatedUnionSynth family (type, empty, SynthCell field, newSynthCell init, cellRegisterDiscriminatedUnion, emitDiscriminatedUnionSynthDecls) mirroring the Map/Record synth; the discriminant rule is unguarded, variant-field rules carry the discriminant-equality guard as a PantRule.guard OpaqueExpr.
- Drain the new synth in cellEmitSynth.
- Add detection + emission unit tests and the fixtures (positive + negative); register the positives in the allowlist; add skipped end-to-end stubs.
- Confirm no emitted-document change (the new path is unconsumed by mapTsType/field access).

## Files to Modify
- tools/ts2pant/src/translate-types.ts (modify): Add `detectDiscriminatedUnion(type, checker)` (structural: a field present on every union member whose type is a distinct literal per member; reads literal values via the StringLiteral/NumberLiteral/BooleanLiteral flags + `.value`; returns the discriminant name + per-variant {literal, fields} or null). Add a `DiscriminatedUnionSynth` type + `emptyDiscriminatedUnionSynth`, a `discriminatedUnionSynth` field on SynthCell + its init in newSynthCell, `cellRegisterDiscriminatedUnion` (dedupe by structural key), and `emitDiscriminatedUnionSynthDecls` (one domain; an unguarded discriminant rule; per-variant field rules guarded by `ast.binop(ast.opEq(), ast.app(ast.var(discRule), [s]), <litSort literal>)`, a field on a subset guarded by the disjunction). Drain it in cellEmitSynth (order: opaque, map, discriminated-union, record). No mapTsType/field-access wiring (dormant).
- tools/ts2pant/tests/translate-types.test.mts (modify): Unit tests: detectDiscriminatedUnion accepts a `{kind:'a';x}|{kind:'b';y}` union and rejects non-discriminated unions (no common field / non-literal discriminant / non-distinct literals); emitDiscriminatedUnionSynthDecls produces the domain + unguarded discriminant rule + guarded variant-field rules. Add skipped stubs (PENDING Patch 2) for the end-to-end field-access emission.
- tools/ts2pant/tests/fixtures/constructs/expressions-discriminated-union.ts (create): Fixtures: discriminant read (`x.kind`), a per-variant field read (`x.r`), a field shared across all variants; plus a negative non-discriminated union (must keep the ambiguous-owner refusal).
- tools/ts2pant/tests/known-typecheck-failures.mts (modify): Register the positive discriminated-union fixtures under a new `discriminated-union` tag; the non-discriminated negative stays under its existing tag.

## Established Precedents

This patch should adopt the following proven techniques rather than rolling its own. Read the references when you need detail on the API shape or invariants they impose. Each entry names a library, algorithm, pattern, paper, RFC, doc, or blog post; the trailing sentence explains how it applies to this specific patch.

- **[algorithm] McCarthy theory of arrays + Dafny-style precondition guards** — https://doi.org/10.1145/321033.321034 — A variant field is a partial function defined only where its discriminant guard holds — modeled (like the Map encoding) as a rule guarded by a predicate, the select-guarded-by-precondition shape. Informs the guarded-rule emission.
- **[paper] Kroening & Strichman, Decision Procedures Ch. 4 (EUF)** — https://www.decision-procedures.org/ — The discriminant and variant-field rules are uninterpreted functions; soundness rests on congruence and the discriminant-literal equalities add only sound facts. No body axioms beyond the guard.

## Gameplan Specification

```
module DU_TAGGED_ENCODING.

> A TypeScript discriminated union — detected structurally (a field
> present on every member whose type is a distinct literal on each, not
> by field name) — is encoded as ONE Pantagruel domain with an unguarded
> discriminant rule and per-variant field rules guarded by the
> discriminant literal (analogous to the Map membership-guarded
> encoding). Field access on a discriminated-union receiver emits the
> corresponding rule, replacing the sum-of-records encoding and the
> ambiguous-owner refusal. Sound by construction (congruence; the guards
> are undischarged until narrowing lands).

Union.
Field.
Rule.

discriminated? u: Union => Bool.
unguarded-discriminant? u: Union => Bool.
field-resolves? f: Field => Bool.
has-body-axiom? r: Rule => Bool.
typechecks? u: Union => Bool.

---

> A discriminated union's discriminant rule is unguarded; the encoding
> carries no body axioms (EUF / congruence only).
all u: Union | discriminated? u -> unguarded-discriminant? u.
all r: Rule | ~(has-body-axiom? r).
> Field access on a discriminated-union receiver resolves to a single
> rule (no ambiguous-owner refusal) and the document type-checks.
all u: Union | discriminated? u -> typechecks? u.
all f: Field | field-resolves? f -> field-resolves? f.
```

## Patch Specification

```
module DU_TAGGED_ENCODING_PATCH_1.

Union.
Rule.

discriminated? u: Union => Bool.
detected-structurally? u: Union => Bool.
emits-domain-and-rules? u: Union => Bool.
has-body-axiom? r: Rule => Bool.

---

> Detection is structural (a field with a distinct literal type on every
> member), never field-name-based. The synth machinery emits one domain
> + discriminant rule + guarded variant-field rules, no body axioms.
> Dormant until patch 2 (unused by mapTsType/field access).
all u: Union | discriminated? u -> detected-structurally? u and emits-domain-and-rules? u.
all r: Rule | ~(has-body-axiom? r).
```

## Implementation Notes

- Detection currently rejects unions that include `null`/`undefined`/`void`; optional/list-lifted DU support is left to the later wiring layer so this patch stays strictly dormant and conservative.
- Numeric discriminant literals are detected, but synth registration only emits guards for non-negative integer literals because the current Pant AST bridge exposes `litNat` but no direct real/negative numeric literal constructor. Non-emittable numeric literal shapes return `null` from registration instead of producing a partial declaration set.
- The synth stores field guards by variant literal key and emits one accessor rule per non-discriminant field, guarded by the OR of all variants where that field exists. This covers subset fields and common fields with the same mechanism; the selected discriminant rule is the only unguarded field rule.
- The constructs snapshot intentionally still shows the old sum-of-records emission and ambiguous-owner behavior. That is the evidence that Patch 1 did not consume the new synth from `mapTsType` or field access.

Spec/Evidence Mapping:
- Structural detection and deterministic tie-break: `detectDiscriminatedUnion` in `tools/ts2pant/src/translate-types.ts`; unit test `detectDiscriminatedUnion accepts structural discriminants and picks sorted first on ties`.
- One domain, unguarded discriminant, guarded variant fields, no body axioms: `cellRegisterDiscriminatedUnion` plus `emitDiscriminatedUnionSynthDecls`; unit test `emitDiscriminatedUnionSynthDecls emits one domain, an unguarded discriminant, and guarded fields`.
- Dormant until Patch 2: no calls were added from `mapTsType` or field-owner resolution; `expressions-discriminated-union.ts` snapshots remain allowlisted and continue to show the pre-existing behavior.
- Required context used: Map guarded-rule emission in `emitSynthDecls`, record synth naming/field-rule emission, `PantRule.guard` rendering in `emit.ts`, and the constructs/dogfood snapshot harness.

### CODE REVIEW

File: tools/ts2pant/src/translate-types.ts [hunk 1]
```diff
@@ -495,6 +496,384 @@ export function emitRecordSynthDecls(
...
+    case "string":
+      return "String";
+    case "number":
+      return strategy.mapNumber();
+    case "boolean":
+      return "Bool";
+    default: {
+      const _exhaustive: never = lit;
+      throw new Error(`Unhandled discriminant literal: ${_exhaustive}`);
+    }
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply tools/ts2pant/src/translate-types.ts:561]: Addressed in commit a90f8b2. The DU literal helper defaults now keep the exhaustive never check but return the never value instead of throwing, so there is no exception path while still satisfying the repo's explicit-default switch lint rule. Verified with `just ts2pant-typecheck`, `just ts2pant-lint-ci`, and `tests/translate-types.test.mts`.
